### PR TITLE
Enhancement of the parameters management for the gstreamer camera 

### PR DESCRIPTION
**Pull request by PIERROOOTT**
Currently, there is one big class for cameras using gstreamer. V4l2 is used to get parameters when it is installed, and parameters are changed by renewing the pipeline which is not very optimal. Moreover, some parameters available with the camera were sometimes missing because not searched.

In this PR, the gstreamer camera class is split in two classes in two different files (1a41b1f), one for Linux if v4l-utils is installed and one for the other operating systems (3ba1c5f). The first one is based on v4l2 to get and set the parameters (6619176, 34d6337, bec4067). The second one is simplified and can just set parameters with the gstreamer pipeline (555f1b7).

### CODE REVIEW

File: src/crappy/camera/gstreamer_camera_v4l2.py [hunk 1]
```diff
@@ -0,0 +1,706 @@
...
+          command = ['v4l2-ctl', '--set-ctrl', name+f'={int(value[0])}']
+        self.log(logging.DEBUG, "Setting "+name+f" with command {command}")
+        run(command, capture_output=True, text=True)
+      else:
+        if self._device is not None:
+          command = ['v4l2-ctl', '-d', self._device, '--set-ctrl',
+                     name+f'={int(value)}']
+        else:
+          command = ['v4l2-ctl', '--set-ctrl', name+f'={int(value)}']
+        self.log(logging.DEBUG, "Setting "+name+f" with command {command}")
```
Comment (WeisLeDocto) [inline src/crappy/camera/gstreamer_camera_v4l2.py:618]: Replace with proper f-string

File: src/crappy/camera/gstreamer_camera_v4l2.py [hunk 2]
```diff
@@ -0,0 +1,706 @@
...
+
+    Returns:
+      The setter function.
+    """
+
+    def setter(value) -> None:
+      """The method to set the value of a setting running v4l2-ctl.
+      """
+
+      if isinstance(value, str):
```
Comment (WeisLeDocto) [inline src/crappy/camera/gstreamer_camera_v4l2.py:604]: Comment to indicate which case this corresponds to

File: src/crappy/camera/gstreamer_camera_v4l2.py [hunk 3]
```diff
@@ -0,0 +1,706 @@
...
+        if self._device is not None:
+          command = ['v4l2-ctl', '-d', self._device, '--set-ctrl',
+                     name+f'={int(value[0])}']
+        else:
+          command = ['v4l2-ctl', '--set-ctrl', name+f'={int(value[0])}']
```
Comment (WeisLeDocto) [inline src/crappy/camera/gstreamer_camera_v4l2.py:609]: Does not work for index > 9

File: src/crappy/camera/gstreamer_camera_v4l2.py [hunk 4]
```diff
@@ -0,0 +1,706 @@
...
+        command = ['v4l2-ctl', '-d', self._device, '--get-ctrl', name]
+      else:
+        command = ['v4l2-ctl', '--get-ctrl', name]
+      try:
+        self.log(logging.DEBUG, "Getting " + name + f" with command {command}")
+        value = run(command, capture_output=True, text=True).stdout.split()[-1]
+        for param in self.parameters:
+          if param.name == name:
+            for option in param.options:
+              if value == option[0]:
```
Comment (WeisLeDocto) [inline src/crappy/camera/gstreamer_camera_v4l2.py:701]: Does not work for index > 9

File: src/crappy/camera/gstreamer_camera_v4l2.py [hunk 5]
```diff
@@ -0,0 +1,706 @@
...
+      """
+
+      # Trying to run v4l2-ctl to get the value
+      if self._device is not None:
+        command = ['v4l2-ctl', '-d', self._device, '--get-ctrl', name]
+      else:
+        command = ['v4l2-ctl', '--get-ctrl', name]
+      try:
+        self.log(logging.DEBUG, "Getting " + name + f" with command {command}")
+        value = run(command, capture_output=True, text=True).stdout.split()[-1]
```
Comment (WeisLeDocto) [inline src/crappy/camera/gstreamer_camera_v4l2.py:697]: Does not work if more than just a number is returned

**Review by WeisLeDocto — CHANGES_REQUESTED** [review]
A few problems with fstrings and indexes.
Also, a bug when setting menu parameters.
Otherwise, everything seems fine !
Thank you Captain

File: src/crappy/camera/gstreamer_camera_basic.py [hunk 6]
```diff
@@ -190,115 +181,12 @@ def open(self,
...
-        if ' ' in self._formats[0]:
-          self.add_choice_setting(name='format',
-                                  choices=tuple(self._formats),
-                                  getter=self._get_format,
-                                  setter=self._set_format)
-        # The size is independent of the format
-        else:
-          self.add_choice_setting(name='format', choices=tuple(self._formats),
-                                  setter=self._set_format)
+      self._formats = ['Default', 'MJPG']
```
[the remark this answers is not in the document]
Comment (PIERROOOTT) [reply src/crappy/camera/gstreamer_camera_basic.py:184]: I will do it in an other branch.

**Pull request merged**